components/board/client.go

```diff
@@ -253,6 +265,125 @@ func (dic *digitalInterruptClient) RemoveCallback(c chan Tick) {
 	panic(errUnimplemented)
 }
 
+func (c *client) StreamTicks(ctx context.Context, interrupts []string, ch chan Tick, extra map[string]interface{}) error {
+	c.streamMu.Lock()
+	defer c.streamMu.Unlock()
+	ext, err := protoutils.StructToStructPb(extra)
+	if err != nil {
+		return err
+	}
+	c.extra = ext
+
+	// Start the stream
+	c.streamRunning = true
+	c.activeBackgroundWorkers.Add(1)
+	ctx, cancel := context.WithCancel(ctx)
+	c.cancelBackgroundWorkers = cancel
+	// Create a background go routine that connects to the server's stream
+	utils.PanicCapturingGo(func() {
+		defer c.activeBackgroundWorkers.Done()
+		c.connectTickStream(ctx, interrupts, ch)
+	})
+	c.mu.RLock()
+	ready := c.streamReady
+	c.mu.RUnlock()
+
+	// wait until the stream is ready to return
+	for !ready {
+		c.mu.RLock()
+		ready = c.streamReady
+		c.mu.RUnlock()
+		// wait for 50 ms before checking again
+		if !utils.SelectContextOrWait(ctx, 50*time.Millisecond) {
+			return ctx.Err()
+		}
+	}
+	return nil
+}
+
+func (c *client) connectTickStream(ctx context.Context, interrupts []string, ch chan Tick) {
+	defer func() {
+		c.streamMu.Lock()
+		defer c.streamMu.Unlock()
+		c.mu.Lock()
+		defer c.mu.Unlock()
+		c.streamCancel = nil
+		c.streamRunning = false
+		c.streamReady = false
+	}()
+
+	streamCtx, cancel := context.WithCancel(ctx)
+	c.streamCancel = cancel
+
+	for {
+		c.mu.Lock()
+		c.streamReady = false
+		c.mu.Unlock()
+		select {
+		case <-ctx.Done():
+			return
+		default:
+		}
+
+		req := &pb.StreamTicksRequest{
+			Name:     c.info.name,
+			PinNames: interrupts,
+		}
+
+		stream, err := c.client.StreamTicks(streamCtx, req)
+		if err != nil {
+			c.logger.CError(ctx, err)
+			if utils.SelectContextOrWait(ctx, 3*time.Second) {
+				continue
```

Comment:
If streaming ticks fails repeatedly (e.g., because we had a typo in a pin name), couldn't we get into an infinite loop?

Reply:
no because if there is an incorrect pin name then the server errors and the context cancels.